tensorboard/data/server/run.rs

```diff
@@ -177,6 +214,18 @@ impl RunLoader {
             }
         }
     }
+
+    fn commit_all(&mut self, run: &Run, commit: &Commit) {
+        let runs = commit.runs.read().expect("acquiring runs lock");
+        let mut run = runs
+            .get(run)
+            .unwrap_or_else(|| panic!("no runs entry for {:?}", run))
+            .write()
+            .expect("acquiring tags lock");
+        for (tag, ts) in &mut self.time_series {
```

Comment:
Probably the best reference for the `iter`/`iter_mut`/`into_iter` names
is [this API guideline entry][api].

As to `iter`/`iter_mut` not having traits: part of the reason that this
isn’t necessary is precisely that `Vec<T>`, `&Vec<T>`, and `&mut Vec<T>`
implement `IntoIter` for items `T`, `&T`, and `&mut T`, respectively, so
if you want to be generic over a container with “something like `iter`”
you can just take `IntoIterator<Item = &T>` and pass `&my_vec`.

The trait also turns out to not be as ergonomic as you might like. Look
at the signature of, say, [`HashSet::iter`]:

```rust
pub fn iter<'a>(&'a self) -> hash_set::Iter<'a, T>;
```

The returned iterator has a lifetime tied to the reference to the set
itself, as it must, since it returns references that point _into_ the
hash set. So your trait would be like:

```rust
trait Iterable<'a> {
    type Item: 'a;
    type Iter: Iterator<Item = Self::Item> + 'a;
    
    fn iter(&'a self) -> Self::Iter;
}
```

…and its implementers like:

```rust
use std::collections::hash_set::{self, HashSet};

impl<'a, K: 'a> Iterable<'a> for HashSet<K> {
    type Item = &'a K;
    type Iter = hash_set::Iter<'a, K>;
    
    fn iter(&'a self) -> Self::Iter {
        self.iter()
    }
}
```

…and its use like:

```rust
fn do_iter<'a, T: 'a, I: Iterable<'a, Item = T>>(x: &'a I) -> () where T: core::fmt::Debug {
    for v in x.iter() {
        println!("{:?}", v);
    }
}

fn main() {
    let h: HashSet<_> = vec!["one", "two"].into_iter().collect();
    do_iter(&h);
}
```

…and imho it’s not clear that you’ve really gained much compared to
using `IntoIterator<Item = &'a T>>`, for which the `do_iter(&h)` call
site would be exactly the same.

[`IntoIterator::into_iter`]: https://doc.rust-lang.org/std/iter/trait.IntoIterator.html#tymethod.into_iter
[`HashSet::iter`]: https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.iter
[self]: https://doc.rust-lang.org/std/iter/trait.IntoIterator.html#impl-IntoIterator-25
[api]: https://rust-lang.github.io/api-guidelines/naming.html#methods-on-collections-that-produce-iterators-follow-iter-iter_mut-into_iter-c-iter

Reply:
Maybe the most explicit way to write it is:

```rust
match &tp.tensor_shape {
    Some(s) if !s.dim.is_empty() => return None,
    _ => (),
}
```

You do get a feel for the `as_ref` dance after a while, but the dance is
still there. I suspect that you’ll prefer this form, so I’ll go ahead
and change it.

Reply:
> Aside: This line led me down a slightly confusing rabbit-hole to understand why `&mut` here has this effect. It seems to be that ultimately it's because there's a convention that containers in the standard library implement `IntoIterator` so that it's specialized depending on whether you have the container itself, an `&` ref, or an `&mut` ref into calling `into_iter()`, `iter()`, or `iter_mut()` respectively (which I guess are also just conventional names rather than being traits in their own right? that was slightly surprising to me).

Docs for this on the `std::iter` module have now landed, as of Rust 1.50.0:
<https://doc.rust-lang.org/std/iter/index.html#iterating-by-reference>

Reply:
Nice!  Thanks again for doing that :)